File: src/visualizer.py
#!/usr/bin/env python
import cv2
import numpy as np
import rospy
from cv2.typing import MatLike
from cv_bridge import CvBridge
from perception_msgs.msg import Polygon
from perception_msgs.srv import (
    visualize_polygon_topic_srv,
    visualize_polygon_topic_srvRequest,
    visualize_polygon_topic_srvResponse,
)
from sensor_msgs.msg import Image

import constants
from utils.camera_topic import CameraTopic


class PolygonRenderer:
    bridge = CvBridge()
    topics = {}

    def __init__(self):
        rospy.init_node(constants.POLYGON_RENDERING_NAME, anonymous=True)
        rospy.Service(
            constants.SERVICE_RENDER_POLYGON_TOPIC,
            visualize_polygon_topic_srv,
            self.service_callback,
        )
        self.publisher = rospy.Publisher(constants.TOPIC_POLYGON_RENDERER, Image)
        self.camera = CameraTopic(constants.PEPPER_FRONT_CAMERA)
        self.sid = self.camera.subscribe(self.camera_callback, wait_turns=1)
        rospy.spin()

    def camera_callback(self, image: MatLike):
        for topic in self.topics:
            if self.topics[topic]["last"] is None:
                continue
            polygons = np.array(self.topics[topic]["last"])
            polygons[::2] *= image.shape[1]
            polygons[1::2] *= image.shape[0]
            polygons = np.round(polygons)

            for i in range(0, len(polygons), 4):
                cv2.line(
                    image,
                    (int(polygons[i]), int(polygons[i + 1])),
                    (int(polygons[i + 2]), int(polygons[i + 3])),
                    color=(0, 255, 0),
                    thickness=2,
                )

        self.publisher.publish(self.bridge.cv2_to_imgmsg(image, "bgr8"))

    def polygon_callback(self, pol: Polygon, topic: str):
        self.topics[topic]["last"] = pol.polygon

    def service_callback(self, request: visualize_polygon_topic_srvRequest):
        available_topics = list(
            map(
                lambda x: x[0],
                filter(
                    lambda x: x[1] == "perception_msgs/Polygon",
                    rospy.get_published_topics(),
                ),
            )
        )
        response = visualize_polygon_topic_srvResponse()

        if request.polygon_topic_name not in available_topics:
            response.visualizing = False
            return response

        sub = rospy.Subscriber(
            request.polygon_topic_name,
            Polygon,
            self.polygon_callback,
            callback_args=request.polygon_topic_name,
        )
        self.topics[request.polygon_topic_name] = {"sub": sub, "last": None}
        response.visualizing = True
        return response


if __name__ == "__main__":
    PolygonRenderer()

File: src/visualizer.cpp
#include <ros/ros.h>
#include <ros/master.h>
#include <sensor_msgs/Image.h>
#include <sensor_msgs/image_encodings.h>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>
#include <perception_msgs/Polygon.h>
#include <perception_msgs/visualize_polygon_topic_srv.h>
#include <unordered_map>
#include <mutex>
#include <vector>
#include <string>
#include <algorithm>
#include <cmath>
#include <boost/bind.hpp>

// constants in the same file
namespace constants {
    const std::string POLYGON_RENDERING_NAME = "vision_utilities_polygon_rendering";
    const std::string SERVICE_RENDER_POLYGON_TOPIC = "/vision_utilities/rendering/visualize_polygon_topic_srv";
    const std::string PEPPER_FRONT_CAMERA = "/robot_toolkit_node/camera/front/image_raw";
    const std::string TOPIC_POLYGON_RENDERER = "/vision_utilities/rendering/polygons";
}

class PolygonRenderer {
public:
    PolygonRenderer(ros::NodeHandle& nh) : nh_(nh) {
        service_ = nh_.advertiseService(constants::SERVICE_RENDER_POLYGON_TOPIC,
                                        &PolygonRenderer::serviceCallback, this);
        cam_sub_ = nh_.subscribe(constants::PEPPER_FRONT_CAMERA, 10,
                                 &PolygonRenderer::cameraCallback, this);
        pub_ = nh_.advertise<sensor_msgs::Image>(constants::TOPIC_POLYGON_RENDERER, 10);
    }

private:
    struct TopicInfo {
        ros::Subscriber sub;
        std::vector<double> last;
        bool has_last = false;
    };

    bool serviceCallback(perception_msgs::visualize_polygon_topic_srv::Request &req,
                         perception_msgs::visualize_polygon_topic_srv::Response &res)
    {
        ros::master::V_TopicInfo master_topics;
        if (!ros::master::getTopics(master_topics)) {
            res.visualizing = false;
            return true;
        }

        bool found = false;
        for (const auto &t : master_topics) {
            if (t.datatype == "perception_msgs/Polygon" && t.name == req.polygon_topic_name) {
                found = true;
                break;
            }
        }

        if (!found) {
            res.visualizing = false;
            return true;
        }

        std::lock_guard<std::mutex> lock(topics_mutex_);
        ros::Subscriber sub = nh_.subscribe<perception_msgs::Polygon>(
            req.polygon_topic_name, 10,
            boost::bind(&PolygonRenderer::polygonCallback, this, _1, req.polygon_topic_name));

        TopicInfo info;
        info.sub = sub;
        info.has_last = false;
        info.last.clear();
        topics_map_[req.polygon_topic_name] = info;

        res.visualizing = true;
        return true;
    }

    void polygonCallback(const perception_msgs::Polygon::ConstPtr &msg, const std::string &topic) {
        std::lock_guard<std::mutex> lock(topics_mutex_);
        TopicInfo &info = topics_map_[topic];
        info.last.clear();
        info.last.reserve(msg->polygon.size());
        for (auto v : msg->polygon) info.last.push_back(static_cast<double>(v));
        info.has_last = !info.last.empty();
    }

    void cameraCallback(const sensor_msgs::ImageConstPtr &msg) {
        cv_bridge::CvImagePtr cv_ptr;
        try {
            cv_ptr = cv_bridge::toCvCopy(msg, sensor_msgs::image_encodings::BGR8);
        } catch (cv_bridge::Exception &e) {
            ROS_ERROR("cv_bridge exception: %s", e.what());
            return;
        }

        cv::Mat image = cv_ptr->image;

        std::lock_guard<std::mutex> lock(topics_mutex_);
        for (auto &kv : topics_map_) {
            const TopicInfo &info = kv.second;
            if (!info.has_last || info.last.empty()) continue;

            std::vector<double> polygons = info.last;
            const int w = image.cols;
            const int h = image.rows;

            for (size_t i = 0; i < polygons.size(); ++i) {
                if ((i % 2) == 0)
                    polygons[i] = std::round(polygons[i] * static_cast<double>(w));
                else
                    polygons[i] = std::round(polygons[i] * static_cast<double>(h));
            }

            for (size_t i = 0; i + 3 < polygons.size(); i += 4) {
                int x1 = static_cast<int>(polygons[i]);
                int y1 = static_cast<int>(polygons[i + 1]);
                int x2 = static_cast<int>(polygons[i + 2]);
                int y2 = static_cast<int>(polygons[i + 3]);
                cv::line(image, cv::Point(x1, y1), cv::Point(x2, y2), cv::Scalar(0, 255, 0), 2);
            }
        }

        sensor_msgs::ImagePtr out_msg =
            cv_bridge::CvImage(msg->header, sensor_msgs::image_encodings::BGR8, image).toImageMsg();
        pub_.publish(out_msg);
    }

    ros::NodeHandle nh_;
    ros::Publisher pub_;
    ros::Subscriber cam_sub_;
    ros::ServiceServer service_;
    std::unordered_map<std::string, TopicInfo> topics_map_;
    std::mutex topics_mutex_;
};

int main(int argc, char** argv) {
    ros::init(argc, argv, constants::POLYGON_RENDERING_NAME, ros::init_options::AnonymousName);
    ros::NodeHandle nh;
    PolygonRenderer node(nh);
    ros::spin();
    return 0;
}
